### ISSUE TITLE
on:change call validation function?

### ISSUE THREAD

**Issue by chinch64** [probablykasper/date-picker-svelte#53]
on:change does not seem to fire when DatePicker is changed or closed.  Is there an alernative way of calling a custom validation function when the user selects a date?

**Comment by probablykasper** [comment]
There's only an `on:select` which is called when the user clicks on a date or presses enter (but _only_ in those cases)

What's your use case?

**Comment by chinch64** [comment]
Good morning Kasper,

Thanks for the reply, it's good to hear from you.

My use case is as follows:

I have multiple date time fields that all need to be independently re-validated when any one of them change. They reflect events that occur during an air ambulance mission, and they must all be in a logical order.

For example, “job assigned” cannot occur after “aircraft takeoff”, so I need an event to fire when either the date or the time is changed on either of these fields.

Similarly, if a "patient arrived at hospital” is entered, then the corresponding “departed scene with patient” must be checked to see if it is present. However, if a patient is not transported, then neither of these fields should be entered.

We have about a dozen of these timestamp event fields that are all interrelated, and the contents of any one event field can affect the error checking of the others, with lots of permutations that reflect our business rules.

Ideally, an on:change event would allow me to call my validateEventFields() function which assesses the whole collection of event fields in the context of each other.

Does that make sense?

Best regards, and thanks once again for your email.

Chris

On 29/04/2023, at 1:56 PM, Kasper ***@***.******@***.***>> wrote:



There's only an on:select which is called when the user clicks on a date or presses enter (but only in those cases)

What's your use case?

**Comment by probablykasper** [comment]
Thanks for explaining. Would it work for you to run your validation whenever the `value` changes?
```svelte
$: validate(value)
```

**Comment by chinch64** [comment]
Yes, I think that would work.

Forgive me, but I am very new to Javascript and Svelte (despite 40 years as a developer) - are you proposing a change to my code using your component as it is, or are you considering a change to your components code to make this possible?






Thanks for explaining. Would it work for you to run your validation whenever the value changes?

$: validate(value)

**Comment by probablykasper** [comment]
That would be a change to your code. I'd suggest reading up on the Svelte reactivity/stores if you're unfamiliar with it.

Having an event for user input would still be a good thing to add though

**Comment by chinch64** [comment]
Thanks, I will do as you suggest. I genuinely appreciate your help.

My only other request would be a height styling in the CSS if that’s not already there - I cannot seem to adjust the height of the input to match the height of my tailwind components. Am I missing something?

On 30/04/2023, at 12:45 PM, Kasper ***@***.******@***.***>> wrote:



That would be a change to your code. I'd suggest reading up on the Svelte reactivity/stores if you're unfamiliar with it.

Having an event for user input would still be a good thing to add though

**Comment by probablykasper** [comment]
Yeah for now you'd have to use custom CSS styling with `:global()`. In the future it would be cool to have an action or something to make custom styles possible

**Comment by chinch64** [comment]
Again forgive me, I’ve been a C# backend developer for the last 20 or so years, so I’m still getting used to parts of this.

I appreciate the help.

Cheers
Chris


On 30/04/2023, at 12:53 PM, Kasper ***@***.******@***.***>> wrote:



Yeah for now you'd have to use custom CSS styling with :global(). In the future it would be cool to have an action or something to make custom styles possible

**Comment by j-norwood-young** [comment]
+1 for on:update. Use case: I need to fire a database update whenever a date is updated.

**Comment by j-norwood-young** [comment]
See https://github.com/probablykasper/date-picker-svelte/pull/55

**Comment by chinch64** [comment]
Hi Kasper,

Thanks for the assist.  I’m having a play with this now, and can see how to get the reactive validation working as well as changing the height and widths of the fields via a combination of :global and css variables as follows:

<script>
import { DateInput } from 'date-picker-svelte'
let mydate = new Date()
$: validate(mydate) // using reactive validation as opposed to event registration

function validate(adate)
{
console.log(adate);
}
</script>

<DateInput bind:value={mydate} />

<style>
:global(input) {
height: 40px; /* this changes the height */
border-radius: 5px; /* this does NOT seem to have any effect */
}

:root {
--date-picker-background: #1b1e27;
--date-picker-foreground: #f7f7f7;
--date-input-width: 200px; /* this changes the width */
}
</style>

However I cannot seem to change the border radius of the child input field. Any suggestions on how I might do this?

Many thanks for any advice you can offer.

Kind regards
Chris


On 30/04/2023, at 12:53 PM, Kasper ***@***.******@***.***>> wrote:



Yeah for now you'd have to use custom CSS styling with :global(). In the future it would be cool to have an action or something to make custom styles possible

**Comment by probablykasper** [comment]
That's a CSS specificity issue. In the devtools we can see why:
<img width="343" alt="image" src="https://github.com/probablykasper/date-picker-svelte/assets/11315492/2bacec58-8344-4b84-8c53-baf5b2eb2017">

So `DateInput.svelte` already has the `border-radius` defined for the component (Svelte generates random classes like `.svelte-1bsxr9h` to scope styles to the current component).

Meanwhile, your `border-radius` is a less specific `:global()` style. A good way to make it more specific would be to wrap it in a `<div>`:
```svelte
<div>
	<DateInput bind:value={mydate} />
</div>

<style>
div :global(input) {
	border-radius: 9px;
}
</style>
```

Which results in this CSS:
<img width="189" alt="image" src="https://github.com/probablykasper/date-picker-svelte/assets/11315492/5c1fd92b-8bc6-4308-b8d8-28056ba5b67a">

**Comment by chinch64** [comment]
That is an insanely useful and educational explanation - particularly on how to interpret what I am seeing in the developer tools. Thank you so much, I genuinely appreciate it,

Best regards
Chris

On 20/05/2023, at 22:40, Kasper ***@***.***> wrote:

﻿

That's a CSS specificity issue. In the devtools we can see why:
[image]<https://user-images.githubusercontent.com/11315492/239682247-2bacec58-8344-4b84-8c53-baf5b2eb2017.png>

So DateInput.svelte already has the border-radius defined for the component (Svelte generates random classes like .svelte-1bsxr9h to scope styles to the current component).

Meanwhile, your border-radius is a less specific :global() style. A good way to make it more specific would be to wrap it in a <div>:

<div>
        <DateInput bind:value={mydate} />
</div>

<style>
div :global(input) {
        border-radius: 9px;
}
</style>

Which results in this CSS:
[image]<https://user-images.githubusercontent.com/11315492/239682488-5c1fd92b-8bc6-4308-b8d8-28056ba5b67a.png>
